### PR TITLE
Billing project ID flag for indexer

### PR DESCRIPTION
**Pull request by alahwa**
- On the deploy side, this allows us to use our service accounts to talk to bigquery tables in other projects
- On the local end, this simplifies our setup by enabling us to use the public platinum_genomes directly instead of copying it into our project (although user now needs to specify a billing project)

Tested with local quickstart + deploy in ah-test-data-explorer

### CODE REVIEW

**Comment by alahwa** [comment]
By the way, I patched in some of your changes since we're overlapping on some stuff here. Let's get your changes in first then I'll merge and then we can look at the finished README (so you can probably ignore the README for now)

File: bigquery/README.md [hunk 1]
```diff
@@ -2,32 +2,35 @@
 
 ### Quickstart
 
-Index a BigQuery table into an Elasticsearch container on your local machine.
+* If `~/.config/gcloud/application_default_credentials.json` doesn't exist,
+create it by running `gcloud auth application-default login`.
+* Providing a GCP Project ID that you can use as the billing project, run the
```
Comment (melissachang) [inline bigquery/README.md]: * Determine the project that will be billed for querying the BigQuery tables. You must have `bigquery.jobs.create` on this project; for example, any project where you are Viewer/Editor/Owner.
* Run indexer: `BILLING_PROJECT_ID=<billing_project_id> docker-compose up --build`

Same below.

File: bigquery/indexer.py [hunk 2]
```diff
@@ -38,6 +38,12 @@ def parse_args():
         type=str,
         help='Directory containing config files. Can be relative or absolute.',
         default=os.environ.get('DATASET_CONFIG_DIR'))
+    parser.add_argument(
+        '--billing_project_id',
+        type=str,
+        help='The ID of the GCP Project to bill, if separate from the project'
+        ' containing the BigQuery table(s)',
+        default=os.environ.get('BILLING_PROJECT_ID'))
```
Comment (melissachang) [inline bigquery/indexer.py]: add required=True

Right now, for local run if I don't pass in BILLING_PROJECT_ID, I get:

indexer_1        |   File "/usr/local/lib/python2.7/site-packages/pandas_gbq/gbq.py", line 819, in read_gbq
indexer_1        |     raise TypeError("Missing required parameter: project_id")
indexer_1        | TypeError: Missing required parameter: project_id
bigquery_indexer_1 exited with code 1

Adding required=True will give a friendly error message.

File: bigquery/deploy/README.md [hunk 3]
```diff
@@ -53,8 +53,7 @@
     ```
 
 * Run indexer on GKE
-  * If you are using the default [platinum_genomes dataset](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery/config/platinum_genomes),
-don't forget to [copy to your project and set project IDs in facet_fields.csv](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery#quickstart).
+  * Unless you are using the platinum genomes sample dataset, [set up the DATASET_CONFIG_DIR](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery).
   * Build, tag, and upload the docker image to GCR:
```
[the remark this answers is not in the document]
Comment (alahwa) [reply bigquery/deploy/README.md]: Done

File: bigquery/deploy/README.md [hunk 4]
```diff
@@ -53,8 +53,7 @@
     ```
 
 * Run indexer on GKE
-  * If you are using the default [platinum_genomes dataset](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery/config/platinum_genomes),
-don't forget to [copy to your project and set project IDs in facet_fields.csv](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery#quickstart).
+  * Unless you are using the platinum genomes sample dataset, [set up the DATASET_CONFIG_DIR](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery).
```
[the remark this answers is not in the document]
Comment (alahwa) [reply bigquery/deploy/README.md]: Done

File: bigquery/docker-compose.yml [hunk 5]
```diff
@@ -11,9 +11,16 @@ services:
       - ELASTICSEARCH_URL=http://elasticsearch:9200
       # For convenience, include platinum_genomes by default.
       - DATASET_CONFIG_DIR=${DATASET_CONFIG_DIR:-dataset_config/platinum_genomes}
+      - BILLING_PROJECT_ID=${GOOGLE_CLOUD_PROJECT_ID}
```
[the remark this answers is not in the document]
Comment (alahwa) [reply bigquery/docker-compose.yml]: Changed everything to `BILLING_PROJECT_ID` as per your suggestion, what do you mean by "taken"? I don't see why we shouldn't make this a consistent var name.

File: bigquery/indexer.py [hunk 6]
```diff
@@ -38,6 +38,12 @@ def parse_args():
         type=str,
         help='Directory containing config files. Can be relative or absolute.',
         default=os.environ.get('DATASET_CONFIG_DIR'))
+    parser.add_argument(
+        '--billing_project_id',
+        type=str,
+        help='The ID of the GCP Project to bill, if separate from the project'
```
[the remark this answers is not in the document]
Comment (alahwa) [reply bigquery/indexer.py]: Done

File: bigquery/indexer.py (hunk 2 above)
Comment (alahwa) [reply bigquery/indexer.py]: Done

File: bigquery/README.md (hunk 1 above)
Comment (alahwa) [reply bigquery/README.md]: Done

File: bigquery/README.md [hunk 7]
```diff
@@ -2,50 +2,50 @@
...
     ```
     docker-compose up -d elasticsearch
     ```
+* Determine the project that will be billed for querying the BigQuery tables.
+Minimally, your account must have `bigquery.jobs.create` permissions on this
+project.
 * Run the indexer:
-`DATASET_CONFIG_DIR=dataset_config/<my dataset> docker-compose up --build indexer`
+  ```
+  BILLING_PROJECT_ID=<billing_project_id> DATASET_CONFIG_DIR=dataset_config/MY_DATASET docker-compose up --build indexer
```
Comment (melissachang) [inline bigquery/README.md]: I would like to keep `<my dataset>`.

I think with `DATASET_CONFIG_DIR=dataset_config/<my dataset>`, it's more clear what you should change, than `DATASET_CONFIG_DIR=dataset_config/MY_DATASET`.

File: bigquery/README.md [hunk 8]
```diff
@@ -2,50 +2,50 @@
...
+  ```
 * List Elasticsearch indices: `http://localhost:9200/_cat/indices?v`  
   View Elasticsearch index: `http://localhost:9200/MY_DATASET/_search?pretty=true`
-* If you want to see local Data Explorer UI, [follow these instructions](https://github.com/DataBiosphere/data-explorer/blob/5441559c57ab7a2e0813e8e4fe7e19a9394f1bdf/README.md#run-local-data-explorer-with-a-specific-dataset).
+* Optionally, [bring up the Data Explorer UI](https://github.com/DataBiosphere/data-explorer).
```
Comment (melissachang) [inline bigquery/README.md]: I would like to still link to https://github.com/DataBiosphere/data-explorer/blob/5441559c57ab7a2e0813e8e4fe7e19a9394f1bdf/README.md#run-local-data-explorer-with-a-specific-dataset

Also, I think "a local Data Explorer UI" sounds a little better.

File: bigquery/deploy/README.md [hunk 9]
```diff
@@ -53,9 +53,10 @@
     ```
 
 * Run indexer on GKE
-  * If you are using the default [platinum_genomes dataset](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery/config/platinum_genomes),
-don't forget to [copy to your project and set project IDs in facet_fields.csv](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery#quickstart).
-  * Build, tag, and upload the docker image to GCR:
+  * Make sure the config in `bigquery/dataset_config/MY_DATASET` is filled out.
```
Comment (melissachang) [inline bigquery/deploy/README.md]: I prefer "config files", can you change it back? "config" isn't standard terminology.

File: bigquery/indexer.py [hunk 10]
```diff
@@ -38,6 +38,13 @@ def parse_args():
         type=str,
         help='Directory containing config files. Can be relative or absolute.',
         default=os.environ.get('DATASET_CONFIG_DIR'))
+    parser.add_argument(
+        '--billing_project_id',
+        type=str,
+        help=
+        'The project to be billed for GCP operations. The account running this script must have bigquery.jobs.create on this project.',
```
Comment (melissachang) [inline bigquery/indexer.py]: I prefer my wording. It makes it clear to the user what is being billed (so they have an idea of what they'll be charged).

This project will be billed for querying the BigQuery tables. The account running this script must have bigquery.jobs.create permission on this project.

File: bigquery/indexer.py [hunk 11]
```diff
@@ -122,13 +130,14 @@ def index_facet_field(es, index_name, primary_key, project_id, dataset_id,
     table_name: BigQuery table name.
     field_name: BigQuery field name.
     readable_field_name: Field name for index and Data Explorer UI
+    billing_project_id: GCP Project ID to bill
```
Comment (melissachang) [inline bigquery/indexer.py]: GCP project ID

"project ID" is what Cloud docs use.

File: bigquery/README.md [hunk 12]
```diff
@@ -2,50 +2,50 @@
...
-  ```
-* Change project ids in `dataset_config/platinum_genomes/facet_fields.csv`
-* If `~/.config/gcloud/application_default_credentials.json` doesn't exist, create it by running `gcloud auth application-default login`
-* `docker-compose up --build`
-* View Elasticsearch index:
- `http://localhost:9200/platinum_genomes/_search?pretty=true`
+* If `~/.config/gcloud/application_default_credentials.json` doesn't exist,
+create it by running `gcloud auth application-default login`.
+* Determine the project that will be billed for querying the BigQuery tables.
+Minimally, your account must have `bigquery.jobs.create` permissions on this
```
[the remark this answers is not in the document]
Comment (alahwa) [reply bigquery/README.md]: Done

File: bigquery/README.md (hunk 7 above)
Comment (alahwa) [reply bigquery/README.md]: Done

File: bigquery/README.md (hunk 8 above)
Comment (alahwa) [reply bigquery/README.md]: Done

File: bigquery/deploy/README.md (hunk 9 above)
Comment (alahwa) [reply bigquery/deploy/README.md]: Done

File: bigquery/indexer.py (hunk 10 above)
Comment (alahwa) [reply bigquery/indexer.py]: Done

File: bigquery/indexer.py (hunk 11 above)
Comment (alahwa) [reply bigquery/indexer.py]: Done

File: bigquery/README.md [hunk 13]
```diff
@@ -2,50 +2,50 @@
...
-  ```
-* Change project ids in `dataset_config/platinum_genomes/facet_fields.csv`
-* If `~/.config/gcloud/application_default_credentials.json` doesn't exist, create it by running `gcloud auth application-default login`
-* `docker-compose up --build`
-* View Elasticsearch index:
- `http://localhost:9200/platinum_genomes/_search?pretty=true`
+* If `~/.config/gcloud/application_default_credentials.json` doesn't exist,
+create it by running `gcloud auth application-default login`.
+* Determine the project that will be billed for querying the BigQuery tables.
+Your account must have `bigquery.jobs.create` permissions this project; this
```
Comment (melissachang) [inline bigquery/README.md]: permission

Same below

File: bigquery/README.md [hunk 14]
```diff
@@ -2,50 +2,50 @@
...
+includes any project where you have the Viewer/Editor/Owner role.
+* Run the indexer:
```
Comment (melissachang) [inline bigquery/README.md]: Change back to no "Run the indexer", this command runs indexer and Elasticsearch

File: bigquery/README.md [hunk 15]
```diff
@@ -2,50 +2,50 @@
...
+  `BILLING_PROJECT_ID=<billing_project_id> docker-compose up --build`
+* View the Elasticsearch index at
```
Comment (melissachang) [inline bigquery/README.md]: Change back to

* View Elasticsearch index:

to be consistent with below

File: bigquery/README.md [hunk 16]
```diff
@@ -2,50 +2,50 @@
...
+ `http://localhost:9200/platinum_genomes/_search?pretty=true`.
 
 If you want to run the Data Explorer UI on this dataset, follow the instructions
-below. Note that you will have to reindex the data into an Elasticsearch
+below. Note that you will have to re-index the data into an Elasticsearch
```
Comment (melissachang) [inline bigquery/README.md]: Change back to reindex, that's the Elasticsearch term.

File: bigquery/README.md [hunk 17]
```diff
@@ -2,50 +2,50 @@
...
     ```
     docker-compose up -d elasticsearch
     ```
+* Determine the project that will be billed for querying the BigQuery tables.
+Your account must have `bigquery.jobs.create` permissions this project; this
+includes any project where you have the Viewer/Editor/Owner role.
 * Run the indexer:
-`DATASET_CONFIG_DIR=dataset_config/<my dataset> docker-compose up --build indexer`
+  ```
+  BILLING_PROJECT_ID=<billing_project_id> DATASET_CONFIG_DIR=dataset_config/<my-dataset> docker-compose up --build indexer
```
Comment (melissachang) [inline bigquery/README.md]: I realized that in data-explorer repo, I've been using <my dataset> with no spaces. Please update this whole PR.

File: bigquery/deploy/README.md [hunk 18]
```diff
@@ -53,9 +53,11 @@
     ```
 
 * Run indexer on GKE
-  * If you are using the default [platinum_genomes dataset](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery/config/platinum_genomes),
-don't forget to [copy to your project and set project IDs in facet_fields.csv](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery#quickstart).
-  * Build, tag, and upload the docker image to GCR:
+  * Make sure the config files in `bigquery/dataset_config/MY_DATASET` are
+  filled out.
+  If you don't yet have config for your dataset, follow the [instructions for local deployment](https://github.com/DataBiosphere/data-explorer-indexers/tree/master/bigquery#index-a-custom-dataset-locally)
```
Comment (melissachang) [inline bigquery/deploy/README.md]: config files

File: bigquery/README.md (hunk 13 above)
Comment (alahwa) [reply bigquery/README.md]: Done

File: bigquery/README.md (hunk 14 above)
Comment (alahwa) [reply bigquery/README.md]: Done

File: bigquery/README.md (hunk 15 above)
Comment (alahwa) [reply bigquery/README.md]: Done

File: bigquery/README.md (hunk 16 above)
Comment (alahwa) [reply bigquery/README.md]: Done

File: bigquery/README.md (hunk 17 above)
Comment (alahwa) [reply bigquery/README.md]: I'm not sure what this comment means

File: bigquery/deploy/README.md (hunk 18 above)
Comment (alahwa) [reply bigquery/deploy/README.md]: Done

File: bigquery/indexer.py [hunk 19]
```diff
@@ -38,6 +38,13 @@ def parse_args():
         type=str,
         help='Directory containing config files. Can be relative or absolute.',
         default=os.environ.get('DATASET_CONFIG_DIR'))
+    parser.add_argument(
+        '--billing_project_id',
+        type=str,
+        help=
+        'The project that will be billed for querying BigQuery tables. The account running this script must have bigquery.jobs.create permissions on this project.',
```
Comment (melissachang) [inline bigquery/indexer.py]: permission

File: bigquery/README.md (hunk 17 above)
Comment (melissachang) [reply bigquery/README.md]: Oops, forgot to put backticks around `<my dataset>`. For this whole PR, replace underscores and dashes with spaces, in `<billing_project_id>` and `<my-dataset>`.

File: bigquery/indexer.py (hunk 19 above)
Comment (alahwa) [reply bigquery/indexer.py]: Done

File: bigquery/README.md (hunk 17 above)
Comment (alahwa) [reply bigquery/README.md]: Done, spaces feels kind of weird, I think underscores are more conventional for this, but it doesn't really matter.

**Pull request merged**